Сделаны все таски 1-8

Xentention opened the pull request.

Xentention committed:
Сделаны все таски 1-8

genroelgvozo commented on src/main/java/tasks/Task1.java:
```diff
@@ -23,6 +24,8 @@ public Task1(PersonService personService) {
-    return Collections.emptyList();
+    return personIds.stream().map((id) -> persons.stream()
+        .filter(person -> Objects.equals(person.getId(), id)).findAny().orElse(null))
+        .filter(Objects::nonNull).collect(Collectors.toList());
```
какая асимптотика? (в задании сверху я прошу указать)

Сразу скажу - можно значительно быстрее, так что надо подумать как не искать персону во всем сете на каждую айдишку

Xentention replied:
Был  O(n^2), поправила на return persons.stream().sorted(Comparator.comparing(person -> personIds.indexOf(person.getId()))).collect(Collectors.toList()), теперь по идее O(n)

P.S. закоммитила исправления по этому и другим комментариям

genroelgvozo replied:
сортировка за O(n)?)
но тут даже к сожалению и n log n не будет, так как каждое сравнение не за константу выполняется

genroelgvozo commented on src/main/java/tasks/Task6.java:
```diff
@@ -19,6 +18,9 @@ public class Task6 {
+    Map<Integer, String> areasNames = areas.stream().collect(Collectors.toMap(Area::getId, Area::getName));
+
+    return persons.stream().flatMap(person -> personAreaIds.getOrDefault(person.getId(), Collections.emptySet())
+        .stream().map(areaId -> person.getFirstName() + " - " + areasNames.get(areaId))).collect(Collectors.toSet());
```
на будущее - в map такой формат лучше выносить в метод, потому что это какая-то отдельная логика, и проще править. А тогда его зона ответственности как-то сформатировать пару персона+арейка, и логично туда всю арейку отправить, пусть он сам решает какие поля ему нужны. А тогда будет логично строить мапу с арейками а не сразу с их именами) А то много правок если окажется что кроме названия, нужен еще госномер арейки вставить в строку

Xentention replied:
Спасибо за замечания! Поправила:)

genroelgvozo commented on src/main/java/tasks/Task8.java:
```diff
@@ -25,63 +21,40 @@ public class Task8 {
+    // собираем коллекцию HashMap, беря за ключ - id персоны, а за значение - имя.
+    // HasMap зачем-то был размером 1
+    return persons.stream()
+        .collect(Collectors.toMap(Person::getId, this::convertPersonToString, (v1, v2) -> v1, HashMap::new));
```
а зачем 4 параметр?

Xentention replied:
И правда, он здесь не нужен, возвращаемое значение же не HashMap, убрала

genroelgvozo replied:
он не нужен не только поэтому) а потому что HashMap и так по умолчанию как коллекция в этом коллекторе

genroelgvozo commented on src/main/java/tasks/Task8.java:
```diff
@@ -25,63 +21,40 @@ public class Task8 {
-    }
-    return has;
+    // disjoint проверяет есть ли у коллекций пересечения
+    return !Collections.disjoint(persons1, persons2);
```
какая асимптотика? и можно ли быстрее?

Xentention replied:
Вроде бы O(n) по меньшей коллекции. Да, думаю, persons1.stream().anyMatch(persons2::contains); должно быть быстрее, ведь проверить наличие пересечений в общем случае быстрее, чем их отсутствие

genroelgvozo replied:
почему линейно? сколько выполняется contains например в ArrayList?
Тут ведь Collection, значит любая коллекция может быть

anyMatch не будет быстрее асимптотически.

genroelgvozo commented on src/main/java/tasks/Task8.java:
```diff
@@ -25,63 +21,40 @@ public class Task8 {
-    count = 0;
-    numbers.filter(num -> num % 2 == 0).forEach(num -> count++);
-    return count;
+    return numbers.filter(num -> num % 2 == 0).count();
```
чето не везде комменты почему поправила, тут есть че порассуждать о исходном коде)

Xentention replied:
count тут лишняя, еще и требует быть effective final. В целом код изначальный код менее читабельный

genroelgvozo commented on src/main/java/tasks/Task8.java:
```diff
@@ -25,63 +21,40 @@ public class Task8 {
-    }
-    persons.remove(0);
-    return persons.stream().map(Person::getFirstName).collect(Collectors.toList());
+    return persons.stream().skip(1).map(Person::getFirstName).collect(Collectors.toList());
```
тут тоже бы комментик чем плох изначальный

Xentention replied:
Не обязательно удалять из коллекции элемент когда можно его просто пропустить, к тому же со skip получается читабельнее и проще рефакторить, если пропустить понадобится больше 1 элемента

Xentention committed:
Учтены комментарии к ПР:
    - В задании 1 теперь нет поиска id по всему списку
    - В задании 8 убрано приведение к HashMap в getPersonNames и изменен Collections.disjoint на stream().anyMatch() в hasSamePersons

Xentention committed:
Учтены комментарии к 6 задаче

genroelgvozo commented on src/main/java/tasks/Task6.java:
```diff
@@ -18,9 +18,14 @@ public class Task6 {
+  }
+
+  public static Map<Integer, Area> mapIdToArea(Collection<Area> areas) {
+    return areas.stream().collect(Collectors.toMap(Area::getId, t -> t));
```
видимо непонятный у меня коммент(
имел в виду другое - то что внутри map - это же формат (в какой вид представить)
и его хорошо выносить в метод, чтобы просто было поправить - через тире там, или двоеточиек, или вообще "Кандидат такой-то в городе таком-то", а не искать внутри сложного кода

но окей)
